### DISCUSSION TITLE
Ability to specify a `defaultSort` on a collection by collection basis

### DISCUSSION THREAD

**Discussion by jmikrut** [payloadcms/payload#2377] [Feature Requests & Ideas]
We should expose a property to allow developers to specify a default `sort` property for each collection, which will be used instead of the built-in Payload default of `-createdAt`.

**Comment by wrobson-lllow** [comment]
Is this referring to the property "defaultSort" on the Collection Config? If so it appears to only accept a name of a field and an optional "-" flag. It would be much more useful if it could also accept a function. For example I'm currently trying to sort a collection of version tags semantically i.e. ["0.0.1", "0.0.2", "0.0.3", ...., "0.0.10", "0.0.11"] but defaultSort is sorting them as ["0.0.1", "0.0.10", "0.0.11", "0.0.2", "0.0.3", etc.]

**Reply by DanRibbens** [reply]
This is an interesting idea. However, since sorting is done from the DB, it might be trickier to implement than simply passing a function I am afraid.

**Reply by wrobson-lllow** [reply]
Any suggestions on how I could implement this? Also is there way to sort the options in a relationship field?

**Reply by DanRibbens** [reply]
There might be a way to customize the sort in MongoDB, but it probably requires a custom aggregation.

A workaround in Payload would be to add `beforeChange` and `afterRead` hooks on your version field to store it in a format that sorts how you expect, while appearing as a semantic version.

For this to work you would just split the string on the `'.'`.
Add 1 to the number (so that 0 can be handled correctly).
Multiply by 1000 so that something like '0.0.90' is less than '0.1.0'

For example:
_Input/output_ => _stored as_
'0.11.1' => '1000120002000',
'0.2.1' =>   '100030002000',

Now when sorted, the DB will be handling it correctly because `1000120002000` > `100030002000`

The afterRead just needs to undo this conversion to appear correct.

You could store this calculated version to sort on field in a computed field that is hidden from the admin and populate it from the user input version field.

I'd like to see us add a feature to specificy the sort order for `hasMany` relationships.

What do you think @wrobson-lllow?

**Reply by wrobson-lllow** [reply]
@DanRibbens this looks like it will work well. I'll be sure to give you any updates.

**Reply by wrobson-lllow** [reply]
@DanRibbens  Creating a hidden "number" field that is set the Number conversion you outlined above on a beforeValidate hook has worked for sorting the tags semantically in the collection. However, I have a global that has a relationship field relating to the tags, where the tags still appear out of order. Is there anyway I can make a relationship field sort the same as the collection to which it is relating.

**Reply by DanRibbens** [reply]
The order should be preserved after retrieval if you store it in the preffered order. 

You could sort the incoming related items on a `beforeChange` field hook. You'll have to use the local api to look them up before returning the ids in the proper order based on the hidden number field you made.

**Reply by wrobson-lllow** [reply]
@DanRibbens  I'm still having trouble getting the records to display in the correct order in a relationship field. The order does not seem to be preserved even after re-seeding the database. Could you give me a little more detail on how to sort the incoming data via the beforeChange field hook?
